### ISSUE TITLE
yaml type errors with kustomize and patches

### ISSUE THREAD

**Issue by snuggie12** [argoproj/argo-cd#6158]
Checklist:

* [/] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [/] I've included steps to reproduce the bug.
* [/] I've pasted the output of `argocd version`.

**Describe the bug**

We recently upgraded from 1.7.12 to 1.8.2 and noticed a weird kustomize error. In `base` a `Deployment` has `.spec.replicas` set to `OVERLAY`. In `dev` we pull in `base` and have a patch setting the replicas to `2`.

When running kustomize we see the following:
```
# base
$ kustomize-3.6.1 build | yq e -j '.spec.replicas' - | jq 'type'
"string"

# dev
$ kustomize-3.6.1 build | yq e -j '.spec.replicas' - | jq 'type'
"number"
```

However, when adding to argo the desired manifest becomes `replicas: '2'` which is a string and not the right type. We get the error:

```
error validating data: ValidationError(Deployment.spec.replicas): invalid type for io.k8s.api.apps.v1.DeploymentSpec.replicas: got "string", expected "integer"
```

This can be fixed by changing `base` to have an integer and can break again by switching back to a string.

This somehow seems like argo breaks up the overlay with a base and since the base is a string the patch must also be a string?

**To Reproduce**

base/deploy.yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: string-test-echo
  namespace: demo
spec:
  replicas: OVERLAY
  template:
    spec:
      containers:
      - image: gcr.io/google_containers/echoserver:1.0
        imagePullPolicy: Always
        name: echo
        ports:
        - containerPort: 8080
          name: main
          protocol: TCP
        resources:
          limits:
            cpu: 20m
            memory: 20Mi
          requests:
            cpu: 10m
            memory: 10Mi
```

dev/deploy.patch.yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: string-test-echo
  namespace: demo
spec:
  replicas: 2
```

Point the argo app at the `dev` directory

**Expected behavior**

Should render with `replicas: 2` instead of `replicas: '2'`

**Version**

```shell
argocd-server: v1.8.2+94017f2
  BuildDate: 2021-01-10T05:40:54Z
  GitCommit: 94017f2c8d97588d4aa2213713a71d51005ed62d
  GitTreeState: clean
  GoVersion: go1.14.12
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: v3.8.1 2020-07-16T00:58:46Z
  Helm Version: v3.4.1+gc4e7485
  Kubectl Version: v1.17.8
  Jsonnet Version: v0.17.0
```

**Comment by jannfis** [comment]
This seems to be a bug specific to Kustomize v3.8.1, which is delivered as part of Argo CD 1.8 release branch. I have tested it locally with the manifests you have provided:

```shell
# Kustomize v3.8.1 is shipped with Argo CD 1.8.x
$ kustomize-3.8.1 version
{Version:kustomize/v3.8.1 GitCommit:0b359d0ef0272e6545eda0e99aacd63aef99c4d0 BuildDate:2020-07-16T00:58:46Z GoOs:linux GoArch:amd64}
$ kustomize-3.8.1 build | yq e -j '.spec.replicas' - | jq 'type'
"string"

# Kustomize 3.9.4 is part of Argo CD 2.0.x
$ kustomize-3.9.4 version
{Version:kustomize/v3.9.4 GitCommit:e41d94ddef8cca7af2709c4104b41c4d43cd236c BuildDate:2021-02-09T19:22:10Z GoOs:linux GoArch:amd64}
$ kustomize-3.9.4 build | yq e -j '.spec.replicas' - | jq 'type'
"number"
```

Unfortunately, this behavior seems to be persistent in the whole `v3.8` branch of Kustomize:

```shell
$ kustomize-3.8.10 version
{Version:kustomize/v3.8.10 GitCommit:1f1304194d24913b8cfe8a026bfcde38d91b6f8c BuildDate:2021-02-07T16:00:36Z GoOs:linux GoArch:amd64}
$ kustomize-3.8.10 build | yq e -j '.spec.replicas' - | jq 'type'
"string"
```

We do not upgrade the version of Kustomize to a new major or minor version within patch releases of Argo CD, we had too many things breaking in the past doing so. 

I guess your options would be to either upgrade to Argo CD v2.0, or to use a custom version of Kustomize with your 1.8 installation.

**Comment by snuggie12** [comment]
Ahh, my mistake. I misread the `8` of the kustomize version on the repo server for a `6`.

We have several versions of kustomize on our repo server so this will not be a problem to workaround. Thanks for looking into it.
